File: q704.py
from typing import List
# Time complexity : O(log N)
# Space complexity : O(1)
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1
        
        if nums[0] > target or nums[-1] < target:
            return -1

        while left <= right:
            pivot = left + (right - left) // 2
            if nums[pivot] == target:
                return pivot
            if target < nums[pivot]:
                right = pivot - 1
            else:
                left = pivot + 1
        return -1


s = Solution()
print(s.search([5], 5)) # 0
print(s.search([-1,0,3,5,9,12], 9)) # 4
print(s.search([-1,0,3,5,9,12], 2)) # -1

File: q704.java
class Solution {
  public int search(int[] nums, int target) {
    int pivot, left = 0, right = nums.length - 1;
    while (left <= right) {
      pivot = left + (right - left) / 2;
      if (nums[pivot] == target) return pivot;
      if (target < nums[pivot]) right = pivot - 1;
      else left = pivot + 1;
    }
    return -1;
  }

  public static void main(String[] args) {
    int[] nums = {-1,0,3,5,9,12};

    Solution s = new Solution();
    System.out.println(s.search(nums, 9)); // => 4
  }
}